lib/winapi/filemanip.c

```diff
@@ -241,6 +242,141 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
+    DWORD bytesRead;
+
+    RtlInitAnsiString(&sourcePath, lpExistingFileName);
+    InitializeObjectAttributes(&objectAttributes, &sourcePath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(
+            &sourceHandle,
+            FILE_GENERIC_READ,
+            &objectAttributes,
+            &ioStatusBlock,
+            FILE_SHARE_READ,
+            FILE_READ_ATTRIBUTES | FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT | FILE_NO_EA_KNOWLEDGE);
+    if (!NT_SUCCESS(status))
+    {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    // Fail overwriting the file if the target == the source. This is
+    // intentionally done after the check for existence of the source
+    // to set the most appropriate LastError value.
+    if (!strcmp(lpExistingFileName, lpNewFileName)) {
+      NtClose(sourceHandle);
+      SetLastError(ERROR_SHARING_VIOLATION);
+      return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+            sourceHandle,
+            &ioStatusBlock,
+            &networkOpenInformation,
+            sizeof(networkOpenInformation),
+            FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status))
+    {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtCreateFile(
+            &targetHandle,  // FileHandle
+            FILE_GENERIC_WRITE,   // DesiredAccess
+            &objectAttributes, // ObjectAttributes
+            &ioStatusBlock,   // IoStatusBlock
+            &networkOpenInformation.AllocationSize,  // AllocationSize
+            networkOpenInformation.FileAttributes,  // FileAttributes
+            FILE_SHARE_READ | FILE_SHARE_WRITE,  // ShareAccess
+            bFailIfExists ? FILE_CREATE : FILE_SUPERSEDE,  // CreateDisposition
+            FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT);  // CreateOptions
+    if (!NT_SUCCESS(status))
+    {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    readBuffer = MmAllocateContiguousMemory(readBufferSize);
+    if (!readBuffer)
+    {
+        NtClose(sourceHandle);
+        NtClose(targetHandle);
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+        return FALSE;
+    }
+
+    while (TRUE)
+    {
+        const BYTE *bufferPos = readBuffer;
+        if (!ReadFile(sourceHandle, readBuffer, readBufferSize, &bytesRead, NULL))
+        {
+            MmFreeContiguousMemory(readBuffer);
+            NtClose(sourceHandle);
+            NtClose(targetHandle);
+            return FALSE;
+        }
+
+        if (!bytesRead)
+        {
+            break;
+        }
+
+        while (bytesRead > 0)
+        {
```

Comment:
Please follow the code style of the file, i.e. `{` on the same line as `)` for `if` and `while`.

Reply:
Fixed, sorry. 

I even had to fight my autoformat to do it incorrectly, not sure what I was looking at for context that made me do it that way :)

Reply:
For minor fixes like this, please try to squash up your changes into the clean commits, so if the review passes, it can be merged immediately without another round-trip.

Reply:
Will do.